Script Brushes

ihm-tswow opened the pull request:
## Features
- TSWoWs "Script Brushes" extension and the example scripts using luajit
- CMake configurations for automatically cloning all the extra dependencies
- Similar configuration for stormlib, so developers no longer need to install it separately.
- Automatically builds the heightmap version of the NoiseTool pulled from the TSWoW repository.
- New documentation for building noggit on windows and linux with luajit support

ihm-tswow committed:
Update build instructions

ihm-tswow committed:
More cleanup

ihm-tswow committed:
Cleanup 3

bloerwald committed:
noggit: mapchunk: use std::fill, document function, highlight maybe in name

bloerwald committed:
noggit: maptile: add documentation

bloerwald commented on src/noggit/World.cpp:
```diff
@@ -2124,6 +2127,67 @@ void World::selectVertices(math::vector_3d const& pos, float radius)
   });
 }
 
+void World::delete_models(std::vector<selection_type>& const types)
```
Commenting because it is an evil typo: `& const` ≠ `const&`. Shame on MSVC compiling this, gcc etc refuse.

bloerwald commented on src/noggit/World.cpp:
```diff
@@ -2124,6 +2127,67 @@ void World::selectVertices(math::vector_3d const& pos, float radius)
+{
+  math::vector_3d pos_min = math::vector_3d(std::min(pos1.x,pos2.x),std::min(pos1.y,pos2.y),std::min(pos1.z,pos2.z));
+  math::vector_3d pos_max = math::vector_3d(std::max(pos1.x,pos2.x),std::max(pos1.y,pos2.y),std::max(pos1.z,pos2.z));
+  _vertex_center_updated = false;
```
these are not set to `true` in the loop below, and the `select_all_chunks_between` version doesn't touch them. Intended?

ihm-tswow replied:
`select_all_chunks_between` selects chunks for the script selections, while this version of selectVertices is just a square version of the other selectVertices, so I think that's intended since the other selectVertices only sets them to false too (`updateVertexCenter`/`vertexBorderChunks` sets them to true, I think that would apply to this one as well).

I can see however that I never wrapped this function, I probably wrote it before I had sketched out our own selection mechanism.

bloerwald committed:
whitespace

bloerwald committed:
noggit: world: fix: &const -> const&

bloerwald committed:
noggit: map_index: explicit return type, move into cpp

bloerwald committed:
whitespace, add _HAS in macro for clarity

bloerwald committed:
whitespace

bloerwald committed:
cmake: add prefix to option variable

bloerwald committed:
cmake: don't add source_group for scripting if disabled

bloerwald committed:
whitespace

bloerwald committed:
cmake: fix: remove stray comma

bloerwald commented on CMakeLists.txt:
```diff
@@ -577,3 +621,113 @@ add_executable (math-matrix_4x4.test test/math/matrix_4x4.cpp)
+
+add_subdirectory(${STORM_SOURCE} ${STORM_BIN} EXCLUDE_FROM_ALL)
+
+add_dependencies(
```
this should™ not be needed given that it links against it below. is there a specific reason to add it or just a copy&paste thing?

bloerwald replied:
i removed it and nobody complained.

bloerwald commented on CMakeLists.txt:
```diff
@@ -577,3 +621,113 @@ add_executable (math-matrix_4x4.test test/math/matrix_4x4.cpp)
+
+# StormLib
+FetchContent_Declare(storm
+  GIT_REPOSITORY https://github.com/ladislav-zezula/StormLib.git
```
Do we want a fixed version for storm? I guess latest is always fine and we didn't require one before?

ihm-tswow replied:
(resolved the other ones to keep discussion here)

Set versions: 955343c4098c85efc16d208441fd6653f80d64ba

- I've set them to the versions I've last used for testing, guess it's good to have it there so people better know how to change it. Some of them don't have tags, but setting it to a commit hash also works. 

- I've restored daScript to point at the upstream again, my version had no differences that matters anymore. 

- For NoiseTool2, I haven't asked the upstream to merge the heightmap view yet, maybe you want to have the fork under wowdev until I get the courage to ask lol. For this reason, it still points to our version.

bloerwald commented on CMakeLists.txt:
```diff
@@ -577,3 +621,113 @@ add_executable (math-matrix_4x4.test test/math/matrix_4x4.cpp)
+
+  # lodepng
+  FetchContent_Declare(lodepng
+    GIT_REPOSITORY https://github.com/lvandeve/lodepng.git
```
specific version?

bloerwald commented on CMakeLists.txt:
```diff
@@ -577,3 +621,113 @@ add_executable (math-matrix_4x4.test test/math/matrix_4x4.cpp)
+
+  # FastNoise2
+  FetchContent_Declare(fastnoise2
+    GIT_REPOSITORY https://github.com/tswow/FastNoise2.git
```
speciic version?

bloerwald commented on CMakeLists.txt:
```diff
@@ -577,3 +621,113 @@ add_executable (math-matrix_4x4.test test/math/matrix_4x4.cpp)
+
+  # daScript
+  FetchContent_Declare(dascript
+    GIT_REPOSITORY https://github.com/ihm-ts/daScript.git
```
specific version?

bloerwald commented on CMakeLists.txt:
```diff
@@ -577,3 +621,113 @@ add_executable (math-matrix_4x4.test test/math/matrix_4x4.cpp)
+  # daScript
+  FetchContent_Declare(dascript
+    GIT_REPOSITORY https://github.com/ihm-ts/daScript.git
+    # hack that ignores daScript submodules (empty string is bugged)
```
care to elaborate? (never used fetchcontent before)

ihm-tswow replied:
The intended way is to set `GIT_SUBMODULES` to an empty string so we skip their submodules (we don't use them and their CMakelists are broken when used as a subdirectory).

There are many versions of cmake (my preinstalled included) that doesn't like empty strings, so the hack is to set it to some source directory that actually does exist, and CMake will skip submodules for that repository: https://gitlab.kitware.com/cmake/cmake/-/issues/20579

bloerwald commented on CMakeLists.txt:
```diff
@@ -577,3 +621,113 @@ add_executable (math-matrix_4x4.test test/math/matrix_4x4.cpp)
+    set(FASTNOISE_BIN ${fastnoise2_BINARY_DIR} CACHE PATH "FastNoise 2 bin files")
+  endif()
+  add_subdirectory(${FASTNOISE_SOURCE} ${FASTNOISE_BIN} EXCLUDE_FROM_ALL)
+  # hack, but something's wrong in FastNoise2s CMakeLists
```
and sadly that repo is not under your control ;)

bloerwald committed:
cmake: whitespace, inline variables

bloerwald committed:
cmake: only add warning flags to noggit targets

bloerwald committed:
cmake: disable warning that triggers error in libstorm

bloerwald commented on .gitignore:
```diff
@@ -0,0 +1 @@
+examples
```
The build process doesn't generate this file, as far as I see. I assume this is a local file you want to hide? I'd very much prefer not adding those repo-wide. 

You can ignore files locally by using `.git/info/exclude` as you would use `.gitignore` or by setting the git config option `core.excludesFile` to some file with your local filter.

Am I wrong in assuming this is local-only?

ihm-tswow replied:
That's odd, f738ea9 still generates that for me. It's daScript spamming out files into `examples/test/unit_test/_aot_generated` in the source root.

bloerwald replied:
I see, I just didn't realize that. We're pointing to your fork anyway, so can we just remove that `UNITIZE_BUILD` call there and get rid of the ignore here?

ihm-tswow replied:
Oh, that was a leftover from when I tried to actually fix the submodules, but we ignore them now instead so the fork has no purpose currently. I guess you could keep your own fork if you want that, prolly shouldn't be on my user account

bloerwald replied:
I added patches to fastnoise and dascript done while fetching them. the one for dascript addresses this and the files should no longer be generated.

bloerwald commented on src/noggit/scripting/scripts/prop_placer.das:
```diff
@@ -0,0 +1,73 @@
+// This file is part of Noggit3, licensed under GNU General Public License (version 3).
```
[on random file in this directory, applies to all] should we move src/noggit/scripting/scripts/ to example/ or alike? I'm not sure users find them deep in here? Alternatively/and, install them?

ihm-tswow replied:
I couldn't get the install script to work, and it seems to reference files that I couldn't find so I didn't touch it. Installing them would be preferrable I guess, but could also move them to some examples directory. Can look into trying to stop daScript from spamming test files to `examples`.

bloerwald replied:
Yeah, the install stuff is broken. We can still move these files in the source directory to be more visible. Not sure if `examples` is the best name (so it doesn't *have* to clash with that weird directory), but probably any top level directory is better than `src/noggit/scripting/scripts` ;) what about just.. `scripts/`?

ihm-tswow replied:
I think `scripts/` is a good name

ihm-tswow replied:
69b31a050a5d24af8d442b93f7df0880c102c569

bloerwald committed:
noggit: mapview: fix: use correct namespace

bloerwald commented on src/noggit/scripting/script_random.cpp:
```diff
@@ -0,0 +1,72 @@
+
+    float rand_float(script_random &rand, float low, float high)
+    {
+// TODO: prolly not a good idea
```
elaborate? 
what about just using the stl random things?

ihm-tswow replied:
~stl randoms (marsenne twister, unless there is also a smaller one that I didn't know about) is very expensive to seed, which is a problem for scripts since many uses will want to seed with local coordinates (prop placer does this, for example). The problem is more that I likely used an incorrect value there and I should've checked it, I will do that just to ensure it doesn't break on edge numbers.~ completely wrong lol, hadn't used it for a long time but there are plenty of good generators. See af0a34618b2da75252600666863591ef8df27d70

bloerwald commented on src/noggit/scripting/script_random.hpp:
```diff
@@ -0,0 +1,30 @@
+      script_random();
+    };
+
+    int rand_int32(script_random &rand, int low, int high);
```
[not specific to this file] there are a lot of `T&` rather than `T` or `T const&` and I can't imagine they are *all* out-parameters. Is there some reason for all these? some weird API by the dependencies?

ihm-tswow replied:
~That's required by daScript, unfortunately. Variables declared by daScript are either done with `var` (non-const) or `let` (const), and unless I did something completely wrong when I tested that, it wouldn't allow me to pass a non-const to a function accepting const (which is _really_ weird from C++ obviously, but I didn't think too much about it and just complied at the time, I can try it again)~ completely wrong and fixed.

bloerwald commented on src/noggit/scripting/script_noggit_module.hpp:
```diff
@@ -0,0 +1,313 @@
+
+CLASS_TEMPLATE(test_class, int, "test_class_int")
+
+int *test_func()
```
leftover?

bloerwald commented on src/noggit/scripting/script_chunk.hpp:
```diff
@@ -0,0 +1,43 @@
+    struct script_chunk
+    {
+      script_chunk(script_selection *sel, MapChunk *chunk);
+      script_chunk() {}
```
fyi: c++11 has ` T() = default;` which does the same as you would expect, and also allows for optimizations as the compiler knows the semantics.

ihm-tswow replied:
8a561731b5f4475379d85d5c55b7d93225fb87f1

bloerwald commented on CMakeLists.txt:
```diff
@@ -388,6 +388,45 @@ set ( math_headers
 
+
+set ( scripting_sources
+    src/noggit/scripting/script_random.cpp
```
they all have the folder in common already. this isn't c, so we can just make it scripting/random etc.

ihm-tswow replied:
I usually do this specifically for scripting APIs because it makes them easier to spot and search for as they tend to overlap with files from other packages, but that's really more a problem with my editor than anything else. I can change them if you still want to.

bloerwald committed:
whitespace

bloerwald commented on src/noggit/scripting/script_image.cpp:
```diff
@@ -0,0 +1,89 @@
+      unsigned error = lodepng::decode(vec, img._width, img._height, path);
+      if (error)
+      {
+        throw script_exception("Failed to load png image with error code:" + error);
```
MSVC seems to not warn on these, but this is really broken! `error` is an integer, so this is "`error bytes into the string "failed to load png…`", not "append integer". 

I have no idea how they don't warn on this.

bloerwald commented on src/noggit/scripting/script_exception.hpp:
```diff
@@ -0,0 +1,20 @@
+    class script_exception : public std::exception
+    {
+    public:
+      script_exception(const char *msg) : 
```
`std::exception` has neither of these ctors. You probably want to either 
- have private member `std::string what` and initialize that plus implement `what()`, or
- just derive from `std::runtime_error` which *has* that ctor.

Again, no idea why the hell this compiles in MSVC…

ihm-tswow replied:
Added correct header for `std::runtime_error`: e4c26d1ef9029beb7c23c9edbe60d9a1f75e44c3

bloerwald commented on src/noggit/scripting/script_filesystem.cpp:
```diff
@@ -0,0 +1,100 @@
+    {
+      if (!fs::exists(path))
+      {
+        throw script_exception("No such file:" + *path);
```
eek, this is double wrong: `"No such file:" + path` didn't compile because `char* + char*` is not a thing (it should be a `std::string` on one side, or use `strcat` (please don't)). `char* + char` is the same as in the other comment though: the first character of the path given will index into the string literal, i.e. this exception will be straight bogus.

bloerwald commented on src/noggit/scripting/script_noise.cpp:
```diff
@@ -0,0 +1,136 @@
+    {
+      _width = width;
+      _height = height;
+      _start_x = _start_x;
```
heh, that one is evil: you self-assigned instead of using the argument

bloerwald commented on src/noggit/scripting/script_filesystem.cpp:
```diff
@@ -0,0 +1,100 @@
+      std::ifstream t(path);
+      std::string str((std::istreambuf_iterator<char>(t)),
+              std::istreambuf_iterator<char>());
+      return str.c_str();
```
This is a *really* good reason to not use `char const*` but just use `std::string` everywhere: The return value of this is a pointer to the `str` variable on the stack. It will be destroyed instantly once this function returns and the return value will point into thin air. If this works right now, it is pure luck (nothing in between calling this function and using the return value changed the memory that was previously used).

I strongly suggest to not use `char const*` anywhere but in *really* optimized code.

bloerwald replied:
I tried to find a solution for this way longer than I wanted. I sadly still don't know how to fix this. The only thing I can guarantee is that this is wrong and **will** break horribly. One can't return a `std::string` or `das::string` here though, so that's not an option either. I guess one needs to do some manual heap allocation to return here?

:bomb: **This will break.**

ihm-tswow replied:
You're completely right, of course. This is the same reason string parameters didn't work, and the solution is the same (script-local heap allocation). I just never got that function to actually break, so I never thought about what it was actually doing.

bloerwald committed:
noggit: scripting: fix: use std::runtime_error instead of std::exception

bloerwald committed:
noggit: scripting: don't operator+ on char const*

bloerwald committed:
noggit: scripting: fix: don't self-assign but use arguments

bloerwald commented on src/noggit/scripting/scripting_tool.hpp:
```diff
@@ -0,0 +1,127 @@
+      bool _last_right = false;
+
+      float _radius = 0;
+      float _speed = 0;
```
unused, leftover or bug?

ihm-tswow replied:
leftover: 9353033f890c9195b541606aebd503524bc4281e

bloerwald committed:
noggit: scripting: ui: layout uses other variable now

bloerwald committed:
noggit: scripting: ui: remove leftover label

bloerwald commented on src/noggit/scripting/scripting_tool.cpp:
```diff
@@ -0,0 +1,720 @@
+
+      if (_string_arrays.find(name) == _string_arrays.end())
+      {
+        auto layout = new QFormLayout(this);
```
this layout is unused but also there is no call to `this->layout()` below. I assume this is some left over and `_script_settings_layout` below is the correct one, this line can just be removed?

ihm-tswow replied:
leftover: ebe26ccf96dfc268ae1f962e157a31149391b595

bloerwald commented on src/noggit/scripting/scripting_tool.hpp:
```diff
@@ -0,0 +1,127 @@
+        World *world,
+        math::vector_3d pos,
+        noggit::camera *cam,
+        float dt,
```
note: this is probably related to the unused `speed`. if removing one, the other should also be removed. or both be used.

ihm-tswow replied:
dt is now a script function, just forgot to add it before: 18e2ba0d5ff6c72fcf9157579944528c2f4fb128

bloerwald commented on src/noggit/scripting/script_noggit_module.hpp:
```diff
@@ -0,0 +1,313 @@
+  }
+};
+
+REGISTER_MODULE(NoggitModule);
```
This one is also an evil one: this macro defines something, not just declares it, which means that there is one definition *per include of this file*. they luckily are all the same so it isn't *that* bad, but this macro really should only ever be used from a .cpp file.

You're using this file only as an include within a single cpp file, so it *is* correct. What people usually do is either
- inline this file to where it is included to not claim it is independent, or
- name it `*.ipp` or alike to mark that it is *not* a normal header that can be freely included.

I went for option 2 here.

ihm-tswow replied:
ipp it is, same with `script_texture_index.hpp`

ihm-tswow replied:
Did the same with script_texture_index: 729c42acd4233cb1c3ee8b27ea608c6e8f1fa7e5

bloerwald commented on src/noggit/scripting/script_loader.cpp:
```diff
@@ -0,0 +1,318 @@
+static bool initialized = false;
+static int cur_script = -1;
+
+// is not allowed to happen in a namespace
```
~~please elaborate. I would imagine this still works in a global anonymous namespace, but has to be in the global one for lookup reasons within the macros?~~The macros are mediocre, I gave up on this and just accepted it being in the global namespace.

bloerwald commented on src/noggit/scripting/script_noggit_module.hpp:
```diff
@@ -0,0 +1,313 @@
+  addAnnotation(make_smart<name##_annotation>(lib));
+
+template <typename T>
+struct test_class
```
also unused?

bloerwald commented on src/noggit/scripting/script_context.cpp:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    script_context::script_context(World *world, math::vector_3d pos, float outer_radius, float inner_radius, noggit::camera *camera, bool alt, bool shift, bool ctrl, bool space)
+      : _world(world), _pos(pos), _camera(camera),
```
fyi: in c++, the members of a class are initialized in exactly the order given in the class declaration, i.e. the header file. there you have world → holding → camera → pos, which means that this initializer list is misleading. in this specific case there is no difference at all since the two variables that are mis-ordered do not depend on any other members. in general, the initializer list shall have the exact same order as the members in the class have (I have no idea why the language allows this even…), because otherwise you may have things like `_pos (pos), _camera (_pos)` which in this case would be broken: `_pos` is only initialized after `_camera` is, so `_camera` is bogus now.

ihm-tswow replied:
fixed since: a9c63a2db563a80e78d224bba9fda0ad40976c2a

bloerwald committed:
noggit: scripting: rename hpp to ipp: not intended for inclusion

bloerwald committed:
noggit: scripting: fix: ctor order

bloerwald committed:
noggit: scripting: fix: use const& instead of &

Likely: there are way more cases where it should be a const&, but this
is the one that diN#t even compile.

bloerwald committed:
noggit: scripting: heap: use anon namespace

bloerwald committed:
cmake: scripting: avoid error with dascript macros

bloerwald committed:
noggit: scripting: remove testing stuff

bloerwald committed:
cmake: add more warnings, would ahve detecetd issues in scripting PR

bloerwald commented on src/noggit/scripting/script_texture_index.hpp:
```diff
@@ -0,0 +1,155 @@
+
+#include <noggit/scripting/script_selection.hpp>
+
+static TextureIndex texture_index[145] = {
```
This file looks auto-generated. How was it done?

ihm-tswow replied:
No idea why I didn't commit it anywhere. This is not checked for correctness at all aside from me looking at the patterns and saying that looks right: https://gist.github.com/ihm-ts/9c8d06898360fc2119fe5fe384634816

bloerwald replied:
So from what I understand this is index in `8*8+9*9` to index in `64*64`?

It also seems that at chunk borders this clamps to within-chunk, i.e. it is always specific to a single chunk. How/are cross chunk iterations possible?

bloerwald commented:
Note: I did an amazing job on the review so far and didn't really read the MR description too closely, so the string/heap/dascript issues were already somewhat known but I still commented on them. I'm still not happy about them after reading over the PR description again :(

ihm-tswow committed:
Fix msvc name clash

ihm-tswow committed:
Fix runtime_error include

ihm-tswow committed:
Fix CMake error when -Wno-missing-prototypes is unsupported

ihm-tswow committed:
Rename script_texture_index.hpp -> script_texture_index.ipp

ihm-tswow committed:
Add const to script api functions

lol

ihm-tswow committed:
Store all script string return values on the heap temporarily

ihm-tswow committed:
Switch to C++11 default constructors

ihm-tswow committed:
Fix script_context member order

ihm-tswow committed:
Fix it for real

ihm-tswow committed:
Fix it actually for real 100%

Time for a break

ihm-tswow committed:
Replace random generator with minstd

ihm-tswow committed:
Remove unused member

ihm-tswow committed:
Add dt to script_context

ihm-tswow committed:
Remove unused variable

ihm-tswow committed:
Rename script_texture_index

ihm-tswow committed:
Improve documentation

ihm-tswow committed:
Merge remote-tracking branch 'noggit3/default' into default

ihm-tswow committed:
Remove extra c_str and remove unecessary variable

bloerwald committed:
cmake: note what the hack is

bloerwald committed:
cmake: fixup: file move

ihm-tswow committed:
Make real random values exclusive upper bound and add exceptions for invalid ranges

ihm-tswow committed:
Merge branch 'default' of https://github.com/tswow/noggit3 into default

bkleiner commented on src/noggit/scripting/script_context.cpp:
```diff
@@ -0,0 +1,38 @@
+
+    void set_ctx(script_context* nctx)
+    {
+      ctx = nctx;
```
ctx never gets free'd. current context only gets released because it runs out of scope in scripting_tool::sendUpdate

bkleiner commented on src/noggit/scripting/scripting_tool.cpp:
```diff
@@ -0,0 +1,711 @@
+    {
+      cur_tool = this;
+      script_context ctx(world, pos_in, brushRadius(), innerRadius(), cam, holding_alt, holding_shift, holding_ctrl, holding_space, dt);
+      set_ctx(&ctx);
```
script_context seems to be only allowed to be != nullptr within the scope of this function.
consider using a RAII pattern similar to the other `scoped_*` contexts noggit already has (opengl)

ihm-tswow committed:
Add exceptions to avoid crashing noggit

bkleiner commented on src/noggit/scripting/script_loader-noggit_module.ipp:
```diff
@@ -0,0 +1,301 @@
+    }                                      \
+    virtual bool isLocal() const override { return true; }             \
+    virtual bool canCopy() const override { return true; }             \
+    virtual bool canMove() const override { return true; }             \
```
All of the classes exposed to daScript are marked Copy and Move-able, while their c++ counterparts are not.
Can you elaborate on the rational here?

ihm-tswow replied:
copypaste from the module tutorial, no benefit to keeping them movable. `canCopy` must be enabled to allow creating variables at all: `var b = create_image()` involves one copy, and daScript does call the copy constructor here. I kept constructors empty from when I was unsure if C++-side refcounting would somehow be messed up in that process, but it doesn't seem like this will matter anymore.

ihm-tswow replied:
Actually, I realized the semantics I had planned are a little messy, but I will come back to this when I figure out the gc

ihm-tswow committed:
Handle json parse errors and create backups

bkleiner commented on src/noggit/scripting/scripting_tool.hpp:
```diff
@@ -0,0 +1,124 @@
+#include <QtWidgets/QGroupBox>
+#include <QtWidgets/QFormLayout>
+#include <QtWidgets/QPushButton.h>
+#include <QtWidgets/QGridLayout.h>
```
Qt's convention is to omit the `.h` suffix. my distribution of Qt seems to enforce this (Arch Linux, Qt 5.15.2) and thus fails to build

Commit:
scripting_tool: follow qt convention for includes

ihm-tswow committed:
Add more descriptive names for numbers

ihm-tswow committed:
Merge branch 'default' of https://github.com/tswow/noggit3 into default

ihm-tswow committed:
Add documentation

bloerwald commented on Scripting.md:
````diff
@@ -0,0 +1,224 @@
+We can register events by special named functions. The below code illustrates the bare minimum required for a simple left-click event with a custom title:
+
+```py
+require noggit
````
can we let that happen automatically, for UX, or does including it do some magic that registers it somewhere and if multiple files import it something breaks?

ihm-tswow replied:
Not sure if it's possible, but I can look into it

ihm-tswow replied:
I haven't been able to find a way to do this (trying to mess with `das::ModuleGroup` doesn't seem to do anything, and needs to happen before `das::compileDaScript`). At the same time my (hacky) vscode plugin does parse imports to know what files to provide autocompletion for, and I could see some files deliberately not importing noggit.

There is no problem with importing native modules from multiple source files even in the same context, i.e:  
`noggit <- fileA.das -> fileB.das -> noggit` 

On a semi-related note, I tried enabling auto-export to remove the need for [export] tags for events, but that breaks something **very** badly and starts corrupting random data like crazy before it crashes. Doesn't seem like we can remove much more boilerplate, unfortunately, but both the `title` and `select` functions are optional which leaves us with:

```
require noggit

[export]
def left_click()
    print("Hello world");
```

bloerwald replied:
Oh well, totally fine. I expected that for the IDE integration but wasn't sure whether that has a `*.spec` glob. Thanks for having a look!

bloerwald commented on Scripting.md:
```diff
@@ -0,0 +1,224 @@
+Noggit API functions should not be called globally in your scripts, but we still made them globally accessible to make them less cumbersome to write.
+The following two rules define where it's safe to call Noggit functions from in your scripts:
+
+- All functions `add_x_param` should be called from the `select` function **only** (or any functions it calls).
```
can we enforce these things by having some scope state within the script context?

ihm-tswow replied:
not sure exactly what you mean by scope state, do you mean a context parameter or just a C++-side state that checks the type of function with every function call?

```
[export]
def select(ctx)
    add_x_param(ctx,....)

def left_click(ctx2)
    print("{pos(ctx2)}")
```

or:

```
void add_x_param()
{
    if(!is_calling_from_select()) throw ...
}
```

bloerwald replied:
I was referring to the second one, checking on the c++ side. I see that you are doing a check in `get_ctx()` so this might already be the case and perfectly fine as is.

ihm-tswow replied:
Most functions end up using the script context to access the world or some parameter there which will throw an error if it's null, but I could add checks to a few of the functions that don't. (and especially the add_parameter stuff). script_context itself is only valid during click events.

bloerwald commented on Scripting.md:
```diff
@@ -0,0 +1,224 @@
+- `title`: Only used to return a display name for this brush in the tool window. It must have the signature `def title(): string` if implemented.
+
+
+### Quickstart for Da syntax and types
```
suggesting to reorder sections as the language is not our feature, the api is, so it should be up front

ihm-tswow replied:
I think I should restructure the document a little differently for that (since users will still be unfamiliar with dascript in 99.99% of the cases), I think it might be better to have one tutorial and one documentation part, probably in separate documents.

bloerwald commented on src/noggit/scripting/script_chunk.cpp:
```diff
@@ -0,0 +1,182 @@
+        throw script_exception("empty texture (in call to chunk_add_texture)");
+      }
+      std::string tex = std::string(texture);
+      if (!boost::starts_with(tex, "tileset/"))
```
not sure if a good idea. we modify user input, and users may not adhere to Blizzard directory structure but invent their own, including not being prefixed with tileset/

ihm-tswow replied:
oh, I thought that was actually a requirement. My bad

ihm-tswow replied:
a25d58ded56cffe6104e3a4d18d408c653b54811

bloerwald commented on src/noggit/scripting/script_model.cpp:
```diff
@@ -0,0 +1,248 @@
+    {
+      if (model._is_wmo)
+      {
+        return;
```
throw instead?

ihm-tswow replied:
Really torn on this one. On one hand, a poorly written script to randomly resize props that breaks as soon as it hits a wmo, and the user isn't the author so they're stuck. On the other hand, users getting very confused not knowing you can't resize wmo's. Should scripts really be what educates users on this?

My reasoning with the exceptions we do throw is that scripts shouldn't ever end up in the situation where they're reading incorrect script parameters (empty strings etc) and try to use them in the program, while situations like these rely so heavily on where in the game world you're testing that it's a mistake I can see many more people make.

Maybe the best thing is renaming the function `resize_if_m2`

bloerwald replied:
I guess a stern warning in the documentation might also be enough. no hard feelings on this one.

ihm-tswow replied:
a74e0df645cb0cc10e065efcfa39725eaf502529

bloerwald committed:
scripting: tool: use scoped lock + c++ mutex

bloerwald committed:
scripting: tool: remove unused variable

bloerwald commented on CMakeLists.txt:
```diff
@@ -577,3 +624,101 @@ add_executable (math-matrix_4x4.test test/math/matrix_4x4.cpp)
+  FetchContent_Declare (dascript
+    GIT_REPOSITORY https://github.com/ihm-ts/daScript.git
+    # don't fetch submodules, see https://gitlab.kitware.com/cmake/cmake/-/issues/20579
+    GIT_SUBMODULES "src"
```
this generates `error: pathspec 'src' did not match any file(s) known to git` with git 2.24 when reconfiguring :(

The cmake bug was apparently fixed in 3.18, which is July 2020. That is recent, but I'd probably prefer forcing people to use that over having this pain of working around a known bug. 

Opinions?

ihm-tswow replied:
I agree, it's better to upgrade

ihm-tswow committed:
Stop assuming tileset texture location

bloerwald committed:
cmake: require cmake 3.18 instead of using buggy workaround

bloerwald committed:
scripting: tool: use "official" include path for json lib

bloerwald committed:
noggit: world_tile_update_queue: fix: let abstract class have virtual dtor

bloerwald committed:
cmake: dont update dependencies on reconfigure of which we use a random version anyway

bloerwald committed:
cmake: patch dependency instead of using .gitignore workaround

bloerwald committed:
cmake: patch dependency instead of adding include dirs a second time

bloerwald committed:
cmake: use include directories on targets instead of a manual list

ihm-tswow committed:
Add caller function parameter to script_exception

ihm-tswow committed:
Add gc functions to scriptloader

ihm-tswow committed:
fix annotationargument

ihm-tswow committed:
Fix table typo

ihm-tswow committed:
scripting: use dascript string heap for generic heap data

scripting: remove filesystem walking (requires rework if we keep it)

scripting: rework noise to not place smart pointers on the heap

scripting: add heap test functions

ihm-tswow committed:
scripting: simplify more constructors

ihm-tswow committed:
scripting: fix settings issues

scripting: settings: fix inner radius write value

scripting: fix json/tool initialization order

ihm-tswow committed:
scripting: rename API functions

scripting: add other missing changes to noggit.spec.das

ihm-tswow committed:
scripting: tool: reset log scroll on errors

ihm-tswow committed:
scripting: remove script_* prefix from script-side names of C++ types

ihm-tswow committed:
scripting: remove script_* prefix from all script type names instead

ihm-tswow committed:
scripting: update scripts to new api

ihm-tswow committed:
scripting: disable autoexport (too buggy)

ihm-tswow committed:
scripting: hackfix: reload on exceptions to avoid crashing noggit

ihm-tswow committed:
scripting: hackfix: allow most float parameters to accept ints too

ihm-tswow committed:
scripting: cleanup: remove namespace hacks and unused macros

ihm-tswow committed:
scripting: documentation: add warnings about trying to scale wmos

ihm-tswow committed:
scripting: move example scripts to ./scripts

ihm-tswow committed:
cmake: set dependency versions

bloerwald committed:
scripting: fix cast (two word types need braces)

bloerwald committed:
scripting: fix: avoid ambiguous reference to random

bloerwald committed:
Merge remote-tracking branch 'origin/default' into default

bloerwald committed:
Merge remote-tracking branch 'origin/default' into default

bloerwald committed:
noggit: scripting: fix: catch by reference

bloerwald committed:
scripting: avoid some conversions and use Qt's functions instead of reimpl

bloerwald committed:
scripting: fix filename in ipp and hide more

bloerwald committed:
scripting: avoid raw pointer

bloerwald committed:
scripting: don't reimplement call_once()

bloerwald committed:
scripting: elaborate on where requirement for scope comes from

bloerwald committed:
scripting: add Loader class

bloerwald committed:
scripting: let cur_script be class state

bloerwald committed:
scripting: avoid global state in loader

bloerwald committed:
scripting: avoid manual filesystem path handling

bloerwald committed:
scripting: avoid usings

bloerwald committed:
scripting: don't explicitly create missing json keys: done automatically

bloerwald committed:
scripting: let context be public

bloerwald committed:
scripting: avoid global state for cur_tool, use context

bloerwald committed:
scripting: let widgets in scripting panel scale horizontally

bloerwald committed:
scripting: let log window use monospaced font

bloerwald committed:
scripting: loader: const correctness

bloerwald committed:
scripting: tool: avoid global loader

bloerwald committed:
dascript: add wrappers that include "only" the required symbol

They still include a lot because that's their structure, but it
includes less and is able to cope with Qt foreach macros (by disabling
them for the specific file. We don't use them anyway.)

bloerwald committed:
noggit: use NOGGIT_LOG instead of Log to avoid clash with daScript

bloerwald committed:
noggit: use c++ range-for instead of Qt foreach

bloerwald committed:
scripting: tool: let Loader be a member

bloerwald committed:
scripting: loader: inline install_modules

bloerwald committed:
scripting: automatically load and save json on ctor/dtor

bloerwald committed:
scripting: tool: let json object be member

bloerwald committed:
scripting: tool: let cur_profile and script_change_mutex be members

bloerwald committed:
scripting: tool: let json accessors be members

bloerwald committed:
scripting: use anonymous namespaces instead of static

bloerwald committed:
whitespace

bloerwald committed:
scripting: model: don't reimplement boost.variant

bloerwald commented:
script functions appear to not set the "adt modified" flags as they e.g. in models directly set rotation variables

bloerwald committed:
scripting: avoid global state for update context

ihm-tswow committed:
scripting: remove daScript from sources

ihm-tswow committed:
scripting: start setting up lua script loader

ihm-tswow committed:
Merge branch 'default' into lua

ihm-tswow committed:
scripting: split up scripting_tool into script_profiles and script_settings

ihm-tswow committed:
Merge remote-tracking branch 'noggit3/default' into default

ihm-tswow committed:
Merge branch 'default' into lua

ihm-tswow committed:
scripting: set up script context, brushes and events

ihm-tswow committed:
scripting: revert click events

ihm-tswow committed:
scripting: register brush and global functions

ihm-tswow committed:
scripting: add vert_iterator class

ihm-tswow committed:
add tex_iterator class

ihm-tswow committed:
scripting: make tex/vert into classes and register them

ihm-tswow committed:
scripting: make chunk into a class and register it

ihm-tswow committed:
scripting: make selection into a class and register it

ihm-tswow committed:
scripting: make random into a class and register it

ihm-tswow committed:
scripting: make noisemap  into a class and register it

ihm-tswow committed:
scripting: make model/model_iterator into classes and register them

ihm-tswow committed:
scripting: remove script_misc

ihm-tswow committed:
scripting: register math functions

ihm-tswow committed:
scripting: remove old loader module

ihm-tswow committed:
scripting: make image into a class and register it

ihm-tswow committed:
scripting: remove script_heap

ihm-tswow committed:
scripting: register filesystem functions

ihm-tswow committed:
scripting: create script_standard_brush class from old brush_* functions

ihm-tswow committed:
scripting: split up script_tex and script_vert

ihm-tswow committed:
scripting: move vert/tex iterators to base files

ihm-tswow committed:
scripting: remove script_include

ihm-tswow committed:
scripting: replace image raw pointer with vector

ihm-tswow committed:
scripting: replace noise raw pointer with vector

ihm-tswow committed:
scripting: replace models raw array with vector

ihm-tswow committed:
scripting: add iterators to selections

- change iterator vector references into smart pointers

ihm-tswow committed:
scripting: remove garbage

ihm-tswow committed:
scripting: fix brush loading and settings issues

ihm-tswow committed:
scripting: add typed function wrappers and modules

ihm-tswow committed:
scripting: small fixes

ihm-tswow committed:
scripting: replace const char* with std::string

ihm-tswow committed:
scripting: fix proper module loading...

ihm-tswow committed:
scripting: add script_objects, tags and change lua_functions

ihm-tswow committed:
scripting: merge lua_state and script_context

ihm-tswow committed:
scripting: fix script_brush style

ihm-tswow committed:
scripting: make script types inherit script_object

ihm-tswow committed:
scripting: cleanup and fix minor issues

ihm-tswow committed:
scripting: bunch of random fixes

ihm-tswow committed:
scripting: fix chunk and model iterators

bloerwald committed:
Merge branch 'default' into lua

bloerwald committed:
cmake: search for lua and fetch-content sol2

Note: no longer enforces luajit but leaves that to the person
building. Might want to fetch-content luajit as well.

bloerwald committed:
-Wreorder

bloerwald committed:
noggit: scripting: fix: &const -> const&

bloerwald committed:
-Wunused

bloerwald committed:
-Wshadow

bloerwald committed:
missing include

bloerwald committed:
noggit: scripting: fix: declare context before use

bloerwald committed:
-Wreturn-type

bloerwald committed:
-Wdefaulted-function-deleted

bloerwald committed:
Merge remote-tracking branch 'upstream/default' into lua

ihm-tswow committed:
scripting: make image individual channels floats

ihm-tswow committed:
noggit: fix shadowed variables

ihm-tswow committed:
cmake: disable shadow errors

(until we have a solution for errors in boost)

ihm-tswow committed:
Merge branch 'lua' into default

ihm-tswow committed:
scripting: fix iterators

ihm-tswow committed:
noggit: move tick events after drawing

allows creating QMessageBoxes inside tick events

ihm-tswow committed:
scripting: add file writing permissions

ihm-tswow committed:
scripting: fix errors in release builds

ihm-tswow committed:
cmake: track own repository for sol2 lua_error fix

ihm-tswow committed:
scripting: (hackfix) add default padding

ihm-tswow committed:
scripting: don't initialize scripts before the scripting tool

ihm-tswow committed:
scripting: make number sliders optional

ihm-tswow committed:
scripting: fix random arguments

ihm-tswow committed:
scripting: add model filename check

ihm-tswow committed:
scripting: fix m2 paste_params

ihm-tswow committed:
scripting: load sol default libraries

ihm-tswow committed:
scripting: add null_tag for empty lines

ihm-tswow committed:
scripting: restore real zero argument

ihm-tswow committed:
scripting: add write permission check when saving images

ihm-tswow committed:
scripting: replace das scripts with lua

ihm-tswow committed:
scripting: add TypeScript api and config files

If you don't want this in upstream Noggit you can remove it, but global.d.ts is pretty useful for generating documentation

ihm-tswow committed:
scripting: replace TypeScript examples with lua

ihm-tswow committed:
scripting: default select_origin zRadius to xRadius

ihm-tswow committed:
scripting: add bool_tag

ihm-tswow committed:
scripting: add documentation and tutorial

ihm-tswow committed:
Update README.md

ihm-tswow committed:
scripting: initialize table before any access

ihm-tswow committed:
scripting: add missing license headers

ihm-tswow committed:
scripting: add security prompts when writing files

ihm-tswow committed:
scripting: clean terminal style after successfully loading all files

ihm-tswow committed:
scripting: handle script folder errors

ihm-tswow committed:
scripting: hackaround: write correct exception to log window

ihm-tswow committed:
Merge remote-tracking branch 'noggit3/default' into default

ihm-tswow committed:
scripting: fix settings layout

ihm-tswow committed:
scripting: fix settings not appearing when creating profile

ihm-tswow committed:
scripting: allow random functions to take two equal numbers

ihm-tswow committed:
noggit: add testing version flag

ihm-tswow committed:
scripting: fix decimal point

ihm-tswow committed:
scripting: fix crash when changing brush setting types

ihm-tswow committed:
scripting: workaround: fix model scaling

(not sure why the scale parameter stopped working)

ihm-tswow committed:
Merge branch 'default' of https://github.com/ihm-tswow/noggit3 into default

ihm-tswow committed:
scripting: fix creating new images

ihm-tswow committed:
scripting: add get_color to verts

ihm-tswow committed:
scripting: add brush radius setters

- read brush radius in events correctly after updates

ihm-tswow committed:
scripting: add effect methods to chunks

ihm-tswow committed:
scripting: add texture/effect fields to clean.lua

ihm-tswow committed:
scripting: add script_image::set_pixel_floats

ihm-tswow committed:
scripting: add new methods to global.d.ts*

ihm-tswow committed:
scripting: add image texture painting

ihm-tswow committed:
scripting: add description tags

ihm-tswow committed:
scripting: fix glitched invisible sliders

ihm-tswow committed:
scripting: fix bool tags

ihm-tswow committed:
scripting: fix and improve prop placer

- fix overlapping model placement

- add circular brush setting

- add setting to remove all models

ihm-tswow committed:
scripting: add missing fields to global.d.ts

ihm-tswow committed:
scripting: add texture_printer script

ihm-tswow committed:
scripting: update api docs

ihm-tswow committed:
scripting: add script docs build instructions

ihm-tswow committed:
scripting: improve documentation for texture_brush

ihm-tswow committed:
scripting: split up description in texture printer

ihm-tswow committed:
scripting: replace iterators with tables

- add script_procedures module to hold paint_texture (moved from tex_iterator)

- add bounds check to tex selections

- make selection noise padding smaller (10 -> 2)

- remove some magic numbers when iterating tex/verts

ihm-tswow committed:
scripting: update documentation

ihm-tswow committed:
Merge remote-tracking branch 'noggit3/default' into default

ihm-tswow committed:
lol

ihm-tswow committed:
scripting: fix vert::textures return type

ihm-tswow committed:
scripting: add windows build instructions

ihm-tswow committed:
scripting: missing include

ihm-tswow committed:
fix: remove unused variable

ihm-tswow committed:
cmake: disable -Werror=shadow check (breaks boost headers)

ihm-tswow committed:
fix: gcc variant ambiguity errors

ihm-tswow committed:
scripting: documentation: update iteration examples

ihm-tswow committed:
scripting: fix: change default arguments to sol2 overloads

- fix scripting on linux (tested on ubuntu 20.1)

- (add overload definitions to the headers to get as close to the real default arguments as possible)

ihm-tswow committed:
documentation: add new linux build instructions

- building with luajit and the new dependencies

sgzapata commented:
Dear colleagues, I am a PhD candidate in software engineering. I ask for your cooperation in answering this survey:
https://forms.gle/QEeoNZqincktPK4E7

IT IS NOT SPAM. **It will only take you 10 minutes.**

Excuse me, **thank you very much for your help**. Best regards.

Sergio Zapata (szapata@iinfo.unsj.edu.ar)
https://www.researchgate.net/profile/Sergio-Zapata-2
Universidad Nacional de San Juan – Argentina
